File: sf_files/TauTriggerSFs/python/SFProvider.py
import ROOT

class SFProvider:
    supported_decay_modes = [ 0, 1, 10, 11 ]

    def __init__(self, input_file, channel, wp):
        root_file = ROOT.TFile(input_file, "READ")
        if root_file.IsZombie():
            raise RuntimeError('tau_trigger::SFProvider: unable to open "{}".'.format(input_file))

        self.eff_data = {}
        self.eff_mc = {}
        self.sf = {}

        histograms = { "data": self.eff_data , "mc": self.eff_mc, "sf": self.sf }
        for dm in SFProvider.supported_decay_modes:
            for type_name, hist_dict in histograms.iteritems():
                hist_name = '{}_{}_{}_dm{}_fitted'.format(type_name, channel, wp, dm)
                hist_dict[dm] = SFProvider._LoadHistogram(root_file, hist_name)
        root_file.Close()

    def getEfficiencyData(self, tau_pt, tau_dm, unc_scale = 0):
        tau_dm = SFProvider._CheckDM(tau_dm)
        return self._FindBinValue(self.eff_data[tau_dm], tau_pt, unc_scale)

    def getEfficiencyMC(self, tau_pt, tau_dm, unc_scale = 0):
        tau_dm = SFProvider._CheckDM(tau_dm)
        return self._FindBinValue(self.eff_mc[tau_dm], tau_pt, unc_scale)

    def getSF(self, tau_pt, tau_dm, unc_scale = 0):
        tau_dm = SFProvider._CheckDM(tau_dm)
        return self._FindBinValue(self.sf[tau_dm], tau_pt, unc_scale)

    @staticmethod
    def _LoadHistogram(file, name):
        hist = file.Get(name)
        if hist is None or type(hist) != ROOT.TH1F:
            raise RuntimeError('tau_trigger::SFProvider: unable to load "{}" from the root file "{}".' \
                               .format(name, file.GetName()))
        cloned_hist = hist.Clone()
        cloned_hist.SetDirectory(0)
        return cloned_hist

    @staticmethod
    def _FindBinValue(hist, x, unc_scale):
        bin = hist.FindFixBin(x)
        bin = min(hist.GetNbinsX(), max(1, bin))
        return hist.GetBinContent(bin) + unc_scale * hist.GetBinError(bin)

    @staticmethod
    def _CheckDM(tau_dm):
        if tau_dm == 2:
            tau_dm = 1
        if tau_dm not in SFProvider.supported_decay_modes:
            raise RuntimeError('tau_trigger::SFProvider: decay mode = {} is not supported.'.format(tau_dm))
        return tau_dm

File: sf_files/TauTriggerSFs/src/SFProvider.cc
#include "TauAnalysisTools/TauTriggerSFs/interface/SFProvider.h"

#include <algorithm>
#include <sstream>

namespace tau_trigger {

const std::set<int> SFProvider::supported_decay_modes = { 0, 1, 10, 11 };

SFProvider::SFProvider(std::string_view input_file, std::string_view channel, std::string_view wp)
{
    TFile root_file(input_file.data(), "READ");
    if(root_file.IsZombie()) {
        std::ostringstream ss;
        ss << "tau_trigger::SFProvider: unable to open \"" << input_file << "\".";
        throw std::runtime_error(ss.str());
    }

    std::map<std::string, std::map<int, std::unique_ptr<TH1F>>*> histograms = {
        { "data", &eff_data }, { "mc", &eff_mc }, { "sf", &sf }
    };
    for(int dm : supported_decay_modes) {
        for(const auto& entry : histograms) {
            std::ostringstream ss_hist_name;
            ss_hist_name << entry.first << "_" << channel << "_" << wp << "_dm" << dm << "_fitted";
            (*entry.second)[dm].reset(LoadHistogram(root_file, ss_hist_name.str()));
        }
    }
}

float SFProvider::getEfficiencyData(float tau_pt, int tau_dm, int unc_scale) const
{
    tau_dm = CheckDM(tau_dm);
    return FindBinValue(*eff_data.at(tau_dm), tau_pt, unc_scale);
}

float SFProvider::getEfficiencyMC(float tau_pt, int tau_dm, int unc_scale) const
{
    tau_dm = CheckDM(tau_dm);
    return FindBinValue(*eff_mc.at(tau_dm), tau_pt, unc_scale);
}

float SFProvider::getSF(float tau_pt, int tau_dm, int unc_scale) const
{
    tau_dm = CheckDM(tau_dm);
    return FindBinValue(*sf.at(tau_dm), tau_pt, unc_scale);
}

TH1F* SFProvider::LoadHistogram(TFile& file, std::string_view name)
{
    TH1F* hist = dynamic_cast<TH1F*>(file.Get(name.data()));
    if(!hist) {
        std::ostringstream ss;
        ss << "tau_trigger::SFProvider: unable to load \"" << name << "\" from the root file \""
           << file.GetName() << "\".";
        throw std::runtime_error(ss.str());
    }
    TH1F* cloned_hist = dynamic_cast<TH1F*>(hist->Clone());
    cloned_hist->SetDirectory(nullptr);
    return cloned_hist;
}

float SFProvider::FindBinValue(const TH1F& hist, float x, int unc_scale)
{
    int bin = hist.FindFixBin(x);
    bin = std::clamp(bin, 1, hist.GetNbinsX());
    return static_cast<float>(hist.GetBinContent(bin) + unc_scale * hist.GetBinError(bin));
}

int SFProvider::CheckDM(int tau_dm)
{
    if(tau_dm == 2) tau_dm = 1;
    if(!supported_decay_modes.count(tau_dm)) {
        std::ostringstream ss;
        ss << "tau_trigger::SFProvider: decay mode = " << tau_dm << " is not supported.";
        throw std::runtime_error(ss.str());
    }
    return tau_dm;
}

} // namespace tau_trigger
